pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	// only https and http are valid URI schemes
+	if uri.Scheme == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI scheme: %s", input)
+	}
+	if uri.Scheme != "https" && uri.Scheme != "http" {
+		return url.URL{}, fmt.Errorf("unsupported URI scheme: %s", uri.Scheme, input)
+	}
+
+	// a host must be specified
+	if uri.Host == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI host: %s", input)
+	}
+
+	// URI user must be empty
+	if uri.User != nil {
+		// strip password for security reasons
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxxx")
+		return url.URL{}, fmt.Errorf("unsupported username/password: %q", uri)
+	}
+
+	return *uri, nil
+}
+
+// UnmashalText turns text into a RegURL, and is a way to unmarshal a string
+// into a url.URL.
+func (r *RegURL) UnmarshalText(text []byte) (err error) {
+	r.URL, err = newRegURL(string(text))
+	return err
+}
+
+// Registry is a generalized type representing a registry host and a registry
+// mirror, whereas the mirror cannot further be mirrored.
+type Registry struct {
+	URL      RegURL     `toml:"url"`
+	Blocked  bool       `toml:"blocked"`
+	Insecure bool       `toml:"insecure"`
+	Search   bool       `toml:"unqualified-search"`
+	Prefix   string     `toml:"prefix"`
+	Mirrors  []Registry `toml:"mirror"`
+}
+
+// String returns host:port/path of the Registry's URL.
+func (r *Registry) String() string {
+	return r.URL.String()
+}
+
+// tomlConfig is the data type used to marshal and unmarshal the toml config.
+type tomlConfig struct {
+	Registries []Registry `toml:"registry"`
+}
+
+// getUnqualifiedSearchRegistries returns all registries that are configured
+// for unqualified image search (i.e., with Registry.Search == true).
+func getUnqualifiedSearchRegistries(registries []Registry) []Registry {
+	unqualified := []Registry{}
+	for _, reg := range registries {
+		if reg.Search {
+			unqualified = append(unqualified, reg)
+		}
+	}
+	return unqualified
+}
+
+// FindRegistries returns all registries matching the specified image reference.
+// In case of an unqualified image reference, all registries that are configured
+// for unqualified image search are returned.  In case of a fully qualified
+// image reference, the registry with the longest prefix is returned.  An empty
+// array is returned in case no matching registry is found.
+func FindRegistries(imgRef reference.Reference, registries []Registry) ([]Registry, error) {
+	domain := reference.Domain(imgRef.(reference.Named))
+	// if no domain is specified, return a list of registries that enabled
+	// unqualified image search
+	if domain == "" {
+		regs := getUnqualifiedSearchRegistries(registries)
+		return regs, nil
+	}
+
+	regs := []Registry{}
+	reg := Registry{}
+	prefixLen := 0
+	for _, r := range registries {
+		if strings.HasPrefix(imgRef.String(), reg.Prefix) {
+			length := len(reg.Prefix)
+			if length > prefixLen {
+				reg = r
+				prefixLen = length
+			}
+		}
+	}
+	if prefixLen != 0 {
+		regs = append(regs, reg)
```

Comment:
(Non-blocking: The `regs` variable seems ultimately unnecessary; it has either zero or one member, so a `return` here would do just as well.)